### DISCUSSION TITLE
Request to search by top level comments only and sort by like count

### DISCUSSION THREAD

**Discussion by fugitivealien** [sonigy/YCS#42] [Ideas]
I'd like to sort search term results by likes, as well as a top level comment only search. Clicking the like sort before or after results just reloads all comments again and by like count.

**Comment by knoajp** [comment]
I made a user script for my preference to order search results by likes count.

```js
// ==UserScript==
// @name         YouTube YCS Sort By Likes
// @description  description
// @version      1
// @match        https://www.youtube.com/watch?v=*
// @run-at       document-idle
// ==/UserScript==

(function(){
  console.log('YouTube YCS Sort By Likes');

  let timer = setInterval(function(){
    const input = document.querySelector('#ycs-input-search');
    const result = document.querySelector('#ycs-search-result');
    if(input === null) return;
    else clearInterval(timer);
    input.addEventListener('keydown', e => {
      if(e.isComposing) return;
      if(e.key === 'Enter'){
        let observer = observe(result, records => {
          observer.disconnect(); // maybe this line should be removed for when it has paginations.
          Array.from(document.querySelectorAll('.ycs-icon-like + .ycs-like-count'))
          .sort((a, b) => parseInt(a.textContent) - parseInt(b.textContent))
          .forEach(e => document.querySelector('#ycs_wrap_comments').prepend(e.parentNode.parentNode.parentNode.parentNode.parentNode));
        });
      }
    });

    const observe = function(element, callback, options = {childList: true, subtree: false, characterData: false, attributes: false, attributeFilter: undefined}){
      let observer = new MutationObserver(callback.bind(element));
      observer.observe(element, options);
      return observer;
    };
  }, 1000);
})();
```

**Reply by fugitivealien** [reply]
Nice, I'll add this in my Tampermonkey install, thx!

**Reply by Letssharearow** [reply]
EDIT: Syntax highlighting

I have slightly adjusted the code to fit my usecase. I want time-stamps and author replies also to sort by likes, so I added a button for the function

```javascript
// ==UserScript==
// @name         YouTube YCS Sort By Likes
// @description  description
// @version      1
// @match        https://www.youtube.com/watch?v=*
// @run-at       document-idle
// ==/UserScript==

(function(){
  console.log('YouTube YCS Sort By Likes');

  let timer = setInterval(function(){

    const searchButton = document.querySelector('#ycs_btn_search');
    const sortButton = searchButton.cloneNode(true);
    sortButton.id = 'ycs_btn_sort_by_likes';
    sortButton.textContent = 'Sort by Likes';
    searchButton.parentNode.insertBefore(sortButton, searchButton.nextSibling);

      function sortResults(){
          Array.from(document.querySelectorAll('.ycs-icon-like + .ycs-like-count'))
              .sort((a, b) => parseInt(a.textContent) - parseInt(b.textContent))
              .forEach(e => document.querySelector('#ycs_wrap_comments').prepend(e.parentNode.parentNode.parentNode.parentNode.parentNode));
      }

    sortButton.addEventListener('click', e => {
        console.log('sorting');
        sortResults()
    });

    const input = document.querySelector('#ycs-input-search');
    const result = document.querySelector('#ycs-search-result');

    if(input === null) return;
    else clearInterval(timer);

    input.addEventListener('keydown', e => {

      if(e.isComposing) return;
      if(e.key === 'Enter'){
        let observer = observe(result, records => {
          observer.disconnect(); // maybe this line should be removed for when it has paginations.
          sortResults();
        });
      }
    });

    const observe = function(element, callback, options = {childList: true, subtree: false, characterData: false, attributes: false, attributeFilter: undefined}){
      let observer = new MutationObserver(callback.bind(element));
      observer.observe(element, options);
      return observer;
    };
  }, 1000);
})();
```

**Reply by NicPalacio** [reply]
Hi guys @knoajp  , @Letssharearow , thank you so much for the script, it has been really useful !.
I have one issues using it and I want to know if you guys know how to fix it (I am not good with Java 😞 ):

 - When using the button "Sorting by Likes" comments with more than 1K of likes are not properly sorted as they appear in the end and I think it has to do with the format Youtube gives to those big numbers, the script is not properly recognizing that K stands for thousand. Do you know how to fix it?. For an especific example there is [https://www.youtube.com/watch?v=qloOyrRUlu4](this) Youtube video  that has multiple comments with more than 1k likes.
![image](https://github.com/sonigy/YCS/assets/134461451/648cff1c-b800-416b-836d-274895dad8e3)

Thank you guys in advance !

**Reply by knoajp** [reply]
Hi @NicPalacio .
That's another i18n issue for sorting by likes count.
(The [K] issue is: https://github.com/sonigy/YCS/issues/16)

In your case, it may be caused by the difference of comma[,] and period[.].
YCS takes care of only 123.4K format, not 123,4K as most Europeans use.

**Reply by Letssharearow** [reply]
@NicPalacio 
I have updated this script recently, but I didn't realise the sorting was bad. As I see it in the code it just takes the number, but because K isnt a number it's undefined and so maybe at the ending. Or it actually takes the 1,2 as a 1.2 which basically is the same xD
Let me update it later and get back to you

Edit: I Also need to find out how to publish scripts, so you can get the updates anytime. And I can actually sync them easily with my devices

**Reply by Letssharearow** [reply]
![image](https://github.com/sonigy/YCS/assets/82836846/1b289397-f0c4-4429-bb4b-58942e98617c)

I dont even have the K, will try to fix it anyways though

Here is the updated Version. I have added a function to get the likesCount from a textContent and some debug values if anything goes wrong. Since I dont have the K I tried using the devTools to add a K to one of the likes and it seems to be working. If there is anything else you'd like to have improved let me know :)
https://greasyfork.org/en/scripts/467013-youtube-ycs-sort-by-likes

**Reply by NicPalacio** [reply]
"It works perfectly! Thank you so much @Letssharearow . I don't want to abuse your kindness, but I wanted to mention a few things I have come across while using the extension. They are minor details but it would be useful if you can handle them:

1.)  I have observed that it only sort the comments that are visible in the default box and from what I have seen it seems that it loads only 200 comments in the view. Therefore, lets say  there are 800 timestamp comments, , it only takes into account the first 200 comments that are currently visible, rather than all of them, to access the additional comments, I need to click 'show more comments' multiple times before using the 'sort by likes' button. 
![image](https://github.com/sonigy/YCS/assets/134461451/92af942a-6442-4d4f-834c-ce11994f751f)


2.)  Another thing that happens sometimes when opening a video, thee Sort by liked button does not appear and I have to refresh the page, I think it has more to do with the Tamper Monkey extension but Im not sure.

Thank you again for the help, have a nice day !

**Reply by NicPalacio** [reply]
> Hi @NicPalacio . That's another i18n issue for sorting by likes count. (The [K] issue is: #16)
> 
> In your case, it may be caused by the difference of comma[,] and period[.]. YCS takes care of only 123.4K format, not 123,4K as most Europeans use.
@knoajp 
Thank you for the answer, in my case it has to do with the script in Tampemonkey, the default Like sort of the extension works perfect.

**Reply by Letssharearow** [reply]
> 1.) it only takes into account the first 200 comments that are currently visible, rather than all of them, to access the additional comments, I need to click 'show more comments' multiple times before using the 'sort by likes' button

I never thought of this. This is a real issue. YCS doesn't load all the comments and I dont think I can modify this. All I can think of is spam-clicking the load more button with a script. I could create another script for that. Like I also want alll the comments and not just random 200 ones. 
Something like "Always load All comments from YCS" that spams click the load button and when everything is loaded, it will notify you in some way maybe.

> 2.) Another thing that happens sometimes when opening a video, thee Sort by liked button does not appear and I have to refresh the page, I think it has more to do with the Tamper Monkey extension but Im not sure.

This should be fixed. This bug occured, because when you click on another video the page isn't loaded again, which means the script doesnt run. I fixed it by adding the sort button when scrolling, so every time you want to scroll to the comments it should appear. Can you confirm that this is also fixed for you? :)

**Reply by NicPalacio** [reply]
@Letssharearow  Thank you, your answers are really helpful. For the default 200 comments I think just do it manually is not that big of a deal but do you think there is a way that the TaperMonkey script considers all the comments loaded the same way the default Like button works ? (it orders all the comments by likes, not only the 200 visible)

And another one weird thing I notice yesterday was that even with your updated script to consider de K notation, sometimes both the default Like button that orders all comments by like, and the 'Sort by Likes' button created with TamperMonkey don't perfectly order the comments, take a look at the image below, I order all comments by like and in the result, 2.7k is listed sometimes properly at the top but other times it is in the middle of lower values, and there is a 2,8k that should be the first one but it is way down. Do you have an idea why this might happen? The original video is ![this one](https://www.youtube.com/watch?v=riPP8K8VafY&t=2556s)  if you want to test if the same thing happens to you. 

![image](https://github.com/sonigy/YCS/assets/134461451/c5acfba6-4eff-4256-8c01-5a855d3737ea)

**Reply by Letssharearow** [reply]
@NicPalacio 
good catch. From my debug I can see that the numbers behind the comma or dot (in my case) are not regarded in the sorting
so 2,5 K and 2,7 K are both 2 000 so they will just appear in random order. parseInt(2,2 K) will take the first numbers until a non-numeric character I assume. And Im just realizing, why would I use parsteInt on a float value, oops

https://greasyfork.org/en/scripts/467013-youtube-ycs-sort-by-likes updated

**Reply by Letssharearow** [reply]
@NicPalacio 
I have created the "loadAllScript" which spam clicks the loadMore button. Works pretty well  for me. Let me know if it works for you too :D
https://greasyfork.org/en/scripts/467366-youtube-ycs-load-all-items

**Comment by NicPalacio** [comment]
Hi again friend @Letssharearow , sorry for not reaching out last week, I was a little busy. Thanks for your fix !. However, I updated the script to the newer version and while I've noticed it improved the sorting, there is still some cases where it doesn't properly organizes the ones with decimals, like the case of the image, do you have any other idea why? in my case I visualize commas for decimals as well,not dots.

![image](https://github.com/sonigy/YCS/assets/134461451/a6bbb9e7-2cd7-49d3-9409-63aeedded1a2)

Link Video: https://www.youtube.com/watch?v=-5zGuFPxW88 searching for comment with "Twitch"

And as for the spam clicker it works perfectly, it automatically charges all comments so the Load more button not even appear jjejej thanks, you rock 👌  .

**Reply by Letssharearow** [reply]
@NicPalacio Interesting. You have a space between the number and the "K" and also a "," instead of a "."   I assume that this is the reason, although my prior testing showed no difference between comma and dot. 
I couldn't find a quick fix
could you enable debug mode (setting it true in the script) and show me the console in the inspect tool (F12 on the youtube video -> console)

And also, are these comments on the top of the list or at the bottom? And are 3,5 K for example ranked higher or lower?

**Reply by NicPalacio** [reply]
Hi, @Letssharearow! I went to TamperMonkey - Settings and changed the Login level to Debug. I'm not sure if that's what you meant. I inspected the console, and here's what it showed:

![image](https://github.com/sonigy/YCS/assets/134461451/bb169b53-2c43-4f47-b308-ed439ead5fc6)


Regarding the positioning of the comments, the ones in the screenshot are some of the most liked, so they appear at the top. However, I haven't observed a clear pattern. Sometimes, a comment with 3.5k likes is correctly placed above a comment with 3k likes, but other times it's  behind a comment with 3.4k likes when it shouldn't be. Another thing I've noticed is that if I press the "Sort by likes" button again, the order changes, but the order error still persists.

**Reply by Letssharearow** [reply]
@NicPalacio Hey. Because they are mostly sorted correctly, I didn't worry too much. 
It actually was the comma, I thought it worked, but maybe my testing was bad. Here's the fixed version now
https://greasyfork.org/en/scripts/467013-youtube-ycs-sort-by-likes

btw. you don't seem to be very good with programming and such, how do you end up on github?

**Reply by NicPalacio** [reply]
@Letssharearow, thank you! Yeah, I don't code, but I work with data, and I'm familiar with git and its use. However, I don't have any projects. I came here because I wanted to talk to the extension developer about the option to sort comments by likes with timestamps and I noticed that in the Github repository the community can ask questions, so I thought I could get help here, and I was right! I found your answer, and it was exactly what I needed. I appreciate you, man. 

Thank you for your help all along!  Let me now if you develop other tools for Youtube, you seee pretty good at it ,for now I won't bother you anymore, haha. 😄 Have a nice week.

**Reply by Letssharearow** [reply]
@NicPalacio you might be interested in this here https://hadzy.com/ it's good for many comments and you can search and sort as well
Im flattered that you think that I am goot at it, because I think many would disagree. I like to be functional not very perfectional. I really liked the input. As far as I can tell these changes that you requested have improved my life as well. 
There are more things that I have done, but if you're looking for good youtube stuff, I would recommend looking into Stylus and AdashimaaTube. It let's you customize youtube. 
I use it to hide shorts, because I dont want to watch shorts
"Hide youtube watched videos" is also a nice chrome extension that removes watched videos for example from your sub-page

**Reply by NicPalacio** [reply]
@Letssharearow 
Great tools buddy! I am trying them out.

It's true that your code might not be perfect, but it works wonders. As a frequent YouTube viewer, it is always very helpful to analyze comments in a more insightful way. This tool you created has been really useful for me. I use it all the time, and it's really handy for finding the most interesting parts of the video and understanding what other people are thinking. It saves me a lot of time and greatly enhances my experience of consuming and enjoying the content.

I am curious ,do you work with software or any code-related field?

**Reply by Letssharearow** [reply]
@NicPalacio thank you. I'd like to continue the convo, but I think GitHub is the wrong place. You can email my public email adress a telegram, discord, what's app, insta or snapchat account if you want to text :)

**Reply by fugitivealien** [reply]
Now that the extension broke due to Youtube API changes, the project got forked over to a new one called "YCS Continued" think we could discuss how to get the sort by likes working over there? Seems I can't hack my way out of a paper bag trying to get it working. Here's the fork link: https://github.com/pc035860/YCS-cont
